assets/js/base/components/quantity-selector/index.tsx

```diff
@@ -78,19 +81,25 @@ const QuantitySelector = ( {
 				className="wc-block-components-quantity-selector__input"
 				disabled={ disabled }
 				type="number"
-				step="1"
-				min="0"
+				step={ step }
+				min={ minimum }
 				value={ quantity }
 				onKeyDown={ quantityInputOnKeyDown }
 				onChange={ ( event ) => {
-					let value =
-						! isNumber( event.target.value ) || ! event.target.value
-							? 0
-							: parseInt( event.target.value, 10 );
+
+					let valueIsNumber = event.target.value && ( isNumber( event.target.value ) || parseInt( event.target.value, 10 ).toString() === event.target.value );
+					let value         = ! valueIsNumber ? quantity : parseInt( event.target.value, 10 );
+
 					if ( hasMaximum ) {
-						value = Math.min( value, maximum );
+						value = Math.min( value, Math.floor( maximum / step ) * step );
```

Comment:
This assumes `maximum` isn't already a multiple of `step`? I can see cases where this might happen, can you add a comment expanding on this logic?

Reply:
> This assumes `maximum` isn't already a multiple of `step`?

It was written to account for that, yes. I'm not sure if this is front-end jurisdiction, and can already see an issue with cases like:

- min = 2
- step = 3
- max = 9

In this case, max should have been 2+3+3=8 -- so if we want to keep this like here it should probably become:

```
value = Math.min( value, Math.floor( maximum / step ) * step + minimum );
```